src/github.com/getlantern/flashlight/util/http.go

```diff
@@ -2,21 +2,138 @@ package util
+// and direct fronted servers in parallel.
+func NewChainedAndFronted() *chainedAndFronted {
+	cf := &chainedAndFronted{}
+	// Just make sure we initialize the atomic value, as otherwise we have to worry
+	// about nil.
+	cf.chainedSucceeded.Store(false)
+	return cf
+}
+
+// ChainedAndFronted fetches HTTP data in parallel using both chained and fronted
+// servers.
+type chainedAndFronted struct {
+	chainedSucceeded atomic.Value
+}
+
+// This method will attempt to execute the specified HTTP request using both
+// chained and fronted servers, simply returning the first response to
+// arrive. Callers MUST use the Lantern-Fronted-URL HTTP header to
+// specify the fronted URL to use.
+func (c *chainedAndFronted) Do(req *http.Request) (*http.Response, error) {
+	frontedUrl := req.Header.Get("Lantern-Fronted-URL")
+	req.Header.Del("Lantern-Fronted-URL")
+
+	if frontedUrl == "" {
+		return nil, errors.New("Callers MUST specify the fronted URL in the Lantern-Fronted-URL header")
+	}
+	responses := make(chan *http.Response, 2)
+	errs := make(chan error, 2)
+
+	doRequest := func(client *http.Client, req *http.Request, chained bool) {
+		if resp, err := client.Do(req); err != nil {
+			log.Errorf("Could not complete request with: %v, %v", frontedUrl, err)
+			errs <- err
+			if chained {
+				c.chainedSucceeded.Store(false)
+			}
+		} else {
+			if chained {
+				c.chainedSucceeded.Store(true)
+			}
+			responses <- resp
+		}
+	}
+
+	go func() {
+		// If chained servers have been succeeding generally, just keep using
+		// them instead of using fronting.
+		if c.chainedSucceeded.Load().(bool) {
+			// This is a bit of a hack to work with the select logic below.
+			errs <- errors.New("Not using fronting since chained previously succeeded")
+			return
+		}
+		client := direct.NewDirectHttpClient()
+		if r, err := http.NewRequest("GET", frontedUrl, nil); err != nil {
+			log.Errorf("Could not create request for: %v, %v", frontedUrl, err)
+			errs <- err
+		} else {
+			doRequest(client, r, false)
+		}
+	}()
+	go func() {
+		if client, err := HTTPClient("", defaultAddr); err != nil {
+			log.Errorf("Could not create HTTP client: %v", err)
+			errs <- err
+		} else {
+			doRequest(client, req, true)
+		}
+	}()
+
+	success := func(resp *http.Response) bool {
+		return resp.StatusCode > 199 && resp.StatusCode < 300
+	}
+
+	for i := 0; i < 2; i++ {
+		select {
+		case resp := <-responses:
+			if i == 1 {
+				log.Debugf("Got second response -- sending")
+				return resp, nil
+			} else if success(resp) {
+				log.Debugf("Got good response")
+				return resp, nil
+			} else {
+				log.Debugf("Got bad first response -- wait for second")
+				_ = resp.Body.Close()
+			}
+		case err := <-errs:
+			log.Debugf("Got an error: %v", err)
+			if i == 1 {
+				return nil, errors.New("All requests errored")
+			}
+		case <-timeout:
+			log.Errorf("Timed out!")
+			return nil, errors.New("Timed out!")
+		}
+	}
+	return nil, errors.New("Reached end")
+}
+
```

Comment:
Yeah that's an interesting idea @fffw. I was also considering using states to clean it up a little bit. I might fiddle a little bit.